pennylane/ops/qubit.py

```diff
@@ -1601,6 +1603,57 @@ def _matrix(cls, *params):
         return U
 
 
+class ControlledQubitUnitary(QubitUnitary):
+    r"""ControlledQubitUnitary(U, control_wires, wires)
+    Apply an arbitrary fixed unitary to ``wires`` with control from the ``control_wires``.
+
+    **Details:**
+
+    * Number of wires: Any (the operation can act on any number of wires)
+    * Number of parameters: 1
+    * Gradient recipe: None
+
+    Args:
+        U (array[complex]): square unitary matrix
+        control_wires (Sequence[int] or int): the control wire(s)
+        wires (Sequence[int] or int): the wire(s) the unitary acts on
+
+    **Example**
+
+    The following shows how a single-qubit unitary can be applied to wire ``2`` with control on
+    both wires ``0`` and ``1``:
+
+    >>> U = np.array([[ 0.94877869,  0.31594146], [-0.31594146,  0.94877869]])
+    >>> qml.ControlledQubitUnitary(U, control_wires=[0, 1], wires=2)
+    """
+    num_params = 1
+    num_wires = AnyWires
+    par_domain = "A"
+    grad_method = None
+
+    def __init__(self, *params, wires=None, do_queue=True, control_wires=None):
+        if control_wires is None:
+            raise ValueError("Must specify control wires")
+
+        wires = Wires(wires)
+        control_wires = Wires(control_wires)
+
+        if Wires.shared_wires([wires, control_wires]):
+            raise ValueError(
+                "The control wires must be different from the wires specified to apply the unitary"
+            )
+
+        wires = control_wires + wires
+
+        U = params[0]
+        padding = 2 ** len(wires) - len(U)
+        CU = block_diag(np.eye(padding), U)
```

Comment:
Wouldn't this cause `CU` to be the same no matter which wires are control and which are target? :thinking:

Reply:
Yes it causes `CU` to always be in the block-diagonal form:
(I, 0)
(0, U)
so that it is control on the first few wires and target on the last few wires.

The trick is that when `CU` is applied by a device, it must satisfy the order of wires specified by `wires`. For example, we might have `control_wires = 1` and `target_wires = [0, 2]` so that the combined (line 1650) `wires = [1, 0, 2]`. The device then knows that it is the `1` wire that corresponds to the top left of the block, and the `[0, 2]` wires that correspond to the `U` of the block. On `default.qubit`, this contraction is performed by `_apply_unitary`, using `tensordot`.

Reply:
Take for example `qml.Toffoli`. If you look at it's matrix representation, it is always the same - we instead infer the control wires by saying they are the first ones specified by the `wires` argument.

Reply:
Ooh, I see. Thanks @trbromley, I missed that completely. It makes a lot of sense. :slightly_smiling_face:

Reply:
It could be worth leaving a comment on this, was also contemplating about it when only looking at the code, the explanation fits very well.